Branch-SDK/src/io/branch/referral/BranchPreinstall.java

```diff
@@ -0,0 +1,117 @@
+package io.branch.referral;
+
+import android.content.Context;
+import android.text.TextUtils;
+import io.branch.referral.Defines.PreinstallKey;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.Iterator;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+/**
+ * Created by --vbajpai on --24/07/2019 at --13:44 for --android-branch-deep-linking-attribution
+ */
+class BranchPreinstall {
+
+    private static final String SYSTEM_PROPERTIES_CLASS_KEY = "android.os.SystemProperties";
+    private static final String BRANCH_PREINSTALL_PROP_KEY = "io.branch.preinstall.apps.path";
+
+    public static void getPreinstallSystemData(Branch branchInstance, Context context) {
+        if (branchInstance != null) {
+            // check if the SystemProperties has the branch file path added
+            String branchFilePath = checkForBranchPreinstallInSystem();
+            if (!TextUtils.isEmpty(branchFilePath)) {
+                // after getting the file path get the file contents
+                readBranchFile(branchFilePath, branchInstance, context);
+            }
+        }
+    }
+
+    private static String checkForBranchPreinstallInSystem() {
+        String path = null;
+        try {
+            path = (String) Class.forName(SYSTEM_PROPERTIES_CLASS_KEY)
+                    .getMethod("get", String.class).invoke(null, BRANCH_PREINSTALL_PROP_KEY);
+        } catch (Exception e) {
+            return null;
+        }
+        return path;
+    }
+
+    private static void readBranchFile(final String branchFilePath, final Branch branchInstance,
+            final Context context) {
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    final StringBuilder branchFileContent = new StringBuilder();
+                    JSONObject branchFileContentJson;
+                    File branchFile = new File(branchFilePath);
+                    BufferedReader br = new BufferedReader(new FileReader(branchFile));
+                    String line;
+
+                    while ((line = br.readLine()) != null) {
+                        branchFileContent.append(line);
+                    }
+                    br.close();
+                    branchFileContentJson = new JSONObject(branchFileContent.toString().trim());
+
+                    if (!TextUtils.isEmpty(branchFileContentJson.toString())) {
```

Comment:
@apeterson-branch I'm checking the empty file data here.

Reply:
This check isn't strictly necessary anymore because it is now handled at a higher level.  It's ok to leave in though as code changes over time and this will make sure another code path doesn't creep in.